src/compiler/checker.ts

```diff
@@ -7908,6 +7921,33 @@ namespace ts {
             return prop.flags & SymbolFlags.Method && find(prop.declarations, decl => isClassLike(decl.parent));
         }
 
+        function getPartialTypeFromFalsyUnion(type: UnionType): Type | undefined {
+            if (type.types.length === 2) {
+                // getFalsyFlagsOfTypes
+                // getTypeFacts
+                const i = Math.max(type.types.indexOf(falseType),

```

Comment:
why not use `removeDefinitelyFalsyTypes` on the union, then check the identity of the output vs. input to know if you should get add `undefined` to the member when you are building the normal spread type.

Reply:
Good idea. That also covers undefined and null, which I forgot.